src/main/java/com/google/sps/servlets/GroceryServlet.java

```diff
@@ -40,19 +40,33 @@
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {      
     UserService userService = UserServiceFactory.getUserService();
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    String userEmail = userService.getCurrentUser().getEmail();
+    String userEmail;
+    try {
+    userEmail = userService.getCurrentUser().getEmail();
+    } catch (NullPointerException e) {
+        response.setContentType("application/text");
+        response.getWriter().println("You must Sign in before using this function");
+        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+        response.sendRedirect("/grocery.html");
```

Comment:
Wouldn't redirecting it cause the error to be dropped by the UI?

Reply:
yes that's handled in my groceryError branch